### PR TITLE
pkg,sorter,processor(ticdc): support graceful close

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

Issue Number: ref #4699 

### What is changed and how it works?

Add Close method to actor, let actor close gracefully.

### Check List 

Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/processor/pipeline/table_actor.go [hunk 1]
```diff
@@ -159,6 +159,11 @@ func NewTableActor(cdcCtx cdcContext.Context,
 	return table, nil
 }
 
+// Close implements Actor interface.
+// TODO: implements table actor stop here.
+func (t *tableActor) Close() {
```
Comment (overvenus) [inline cdc/processor/pipeline/table_actor.go:164]: PTAL @sdojjy

**Comment by 3AceShowHand** [comment]
can you add more description about the PR, why it can solve the problem that `some captures' sink don't work`

File: pkg/actor/actor.go [hunk 2]
```diff
@@ -41,17 +42,26 @@ type Actor interface {
 	// We choose message to have a concrete type instead of an interface to save
 	// memory allocation.
 	Poll(ctx context.Context, msgs []message.Message) (running bool)
+
+	// Close is called after Poll retrurns false,
```
Comment (3AceShowHand) [inline pkg/actor/actor.go:46]: `retrurns` -> `return`

File: pkg/actor/system.go [hunk 3]
```diff
@@ -68,18 +90,58 @@ func (p *proc) batchReceiveMsgs(batchMsg []message.Message) int {
...
+		&p.state, uint64(procStateMailboxClosed), uint64(procStateClosed)) {
+		p.actor.Close()
+	}
+}
+
+type readyState int
+
+const (
+	// Running state, system needs to poll actors.
+	readyStateRunning readyState = 0
```
Comment (3AceShowHand) [inline pkg/actor/system.go:138]: how about `readyStateRunning readyState = itoa`, and then ignore the following enum number.

**Review by 3AceShowHand — APPROVED** [review]
Rest LGTM

**Comment by overvenus** [comment]
> can you add more description about the PR, why it can solve the problem that `some captures' sink don't work`

The issue will be fixed in a later PR. This PR provides graceful close mechanism which is required by the fix.

**Comment by overvenus** [comment]
> > can you add more description about the PR, why it can solve the problem that `some captures' sink don't work`
> 
> The issue will be fixed in a later PR. This PR provides graceful close mechanism which is required by the fix.

See PR https://github.com/pingcap/tiflow/pull/4831

File: cdc/processor/pipeline/table_actor.go (hunk 1 above)
Comment (sdojjy) [reply cdc/processor/pipeline/table_actor.go:164]: ok, I will fire a pr after this is merged

File: pkg/actor/system.go [hunk 4]
```diff
@@ -533,6 +605,12 @@ func (s *System) poll(ctx context.Context, id int) {
 					s.handleFatal(
 						"try to remove non-existent actor", p.mb.ID())
 				}
+
+				// Drop all remaining messages.
+				if remainMsgCount != 0 {
+					// TODO: we may need to releaes resources hold by messages.
```
Comment (sdojjy) [inline pkg/actor/system.go:611]:
```suggestion
					// TODO: we may need to release resources hold by messages.
```

File: pkg/actor/actor.go [hunk 5]
```diff
@@ -41,17 +42,26 @@ type Actor interface {
 	// We choose message to have a concrete type instead of an interface to save
 	// memory allocation.
 	Poll(ctx context.Context, msgs []message.Message) (running bool)
+
+	// Close is called after Poll returns false,
+	// or actor system is stopping and all message has been received.
+	//
+	// Close must be idempotent and nonblocking.
+	Close()
```
Comment (sdojjy) [inline pkg/actor/actor.go:50]: 1.  for normal cases, resources can be released before `Poll` returns `false`.
2. The name `Close` confused me,  it sounds like we can Close the actor directly, but it's just a callback.

File: pkg/actor/system.go [hunk 6]
```diff
@@ -400,13 +459,21 @@ func (s *System) Start(ctx context.Context) {
 }
 
 // Stop the system, cancels all actors. It should be called after Start.
+// Messages sent before this call will be receive by actors.
```
Comment (sdojjy) [inline pkg/actor/system.go:462]: seems we need `StopNow`, sometimes we need it waste of time to send amessages if the whole system is stopped

**Review by sdojjy — APPROVED** [review]

**Pull request merged**